robots/bebop.h

```diff
@@ -0,0 +1,1188 @@
+ * from the drone.
+ */
+inline void
+Bebop::addEventHandlers()
+{
+    checkError(ARCONTROLLER_Device_AddStateChangedCallback(
+            m_Device.get(), stateChanged, this));
+    checkError(ARCONTROLLER_Device_AddCommandReceivedCallback(
+            m_Device.get(), commandReceived, this));
+}
+
+/*
+ * Invoked by commandReceived().
+ *
+ * Prints the battery state whenever it changes.
+ */
+inline void
+Bebop::onBatteryChanged(const ARCONTROLLER_DICTIONARY_ELEMENT_t *dict)
+{
+    // which command was received?
+    ARCONTROLLER_DICTIONARY_ELEMENT_t *elem = nullptr;
+    HASH_FIND_STR(dict, ARCONTROLLER_DICTIONARY_SINGLE_KEY, elem);
+    if (!elem) {
+        return;
+    }
+
+    // get associated value
+    ARCONTROLLER_DICTIONARY_ARG_t *val = nullptr;
+    HASH_FIND_STR(
+            elem->arguments,
+            ARCONTROLLER_DICTIONARY_KEY_COMMON_COMMONSTATE_BATTERYSTATECHANGED_PERCENT,
+            val);
+
+    if (val) {
+        // print battery status
+        std::cout << "Battery: " << (int) val->value.U8 << "%" << std::endl;
+    }
+}
+
+/*
+ * Empty function used to suppress default ARSDK console messages.
+ */
+int
+Bebop::printCallback(eARSAL_PRINT_LEVEL level,
+                     const char *tag,
+                     const char *format,
+                     va_list va)
+{
+    // do nothing
+    return 0;
+}
+
+/*
+ * Invoked when the drone's connection state changes.
+ */
+void
+Bebop::stateChanged(eARCONTROLLER_DEVICE_STATE newstate,
+                    eARCONTROLLER_ERROR err,
+                    void *data)
+{
+    Bebop *bebop = reinterpret_cast<Bebop *>(data);
+    bebop->m_Semaphore.notify(); // trigger semaphore used by get_state_update()
+
+    switch (newstate) {
+    case ARCONTROLLER_DEVICE_STATE_STOPPED:
+        std::cout << "Drone stopped" << std::endl;
+        break;
+    case ARCONTROLLER_DEVICE_STATE_STARTING:
+        std::cout << "Drone starting..." << std::endl;
+        break;
+    case ARCONTROLLER_DEVICE_STATE_RUNNING:
+        std::cout << "Drone started" << std::endl;
+        break;
+    case ARCONTROLLER_DEVICE_STATE_PAUSED:
+        std::cout << "Drone is paused" << std::endl;
+        break;
+    case ARCONTROLLER_DEVICE_STATE_STOPPING:
+        std::cout << "Drone stopping..." << std::endl;
+        break;
+    default:
+        std::cerr << "Unknown state!" << std::endl;
+    }
+}
+
+#define MAX_SPEED_CHANGED(NAME, KEY)                                         \
+    {                                                                        \
+        bebop->m_##NAME##Limits.onChanged(                                   \
+                dict,                                                        \
+                ARCONTROLLER_DICTIONARY_KEY_ARDRONE3_##KEY##CHANGED_CURRENT, \
+                ARCONTROLLER_DICTIONARY_KEY_ARDRONE3_##KEY##CHANGED_MIN,     \
+                ARCONTROLLER_DICTIONARY_KEY_ARDRONE3_##KEY##CHANGED_MAX);    \
+    }
```

Comment:
This is the sort of use of macros that I totally approve of but I think this is definitely the example of stuff that would be better off out of harm/#include's way in an implementation cc file.

Reply:
Good shout.